File: python/completed/euler0054.py
###
#
# Created by Holden on 8/25/2016
#
# SOLVED on 8/30/2016
#
# Problem 54 - Poker Hands
#
###

class Card:
    
    faceValues = {"2": 2,
                  "3": 3,
                  "4": 4,
                  "5": 5,
                  "6": 6,
                  "7": 7,
                  "8": 8,
                  "9": 9,
                  "T": 10,
                  "J": 11,
                  "Q": 12,
                  "K": 13,
                  "A": 14}
    
    def __init__(self, card):
        self._value = card[0]
        self._suit = card[1]
        self._numericalValue = self.faceValues[self._value]
        
    def __eq__(self, other):
        return self._numericalValue == other._numericalValue
        
    def __lt__(self, other):
        return self._numericalValue < other._numericalValue
    
    def __gt__(self, other):
        return self._numericalValue > other._numericalValue
    
    def __repr__(self):
        return self._value + self._suit
    
    def getValue(self):
        return self._value
    
    def getSuit(self):
        return self._suit
    
    def getNumericalValue(self):
        return self._numericalValue

class Hand:
    
    handValues = {"High Card": 0,
                  "One Pair": 1,
                  "Two Pairs": 2,
                  "Three of a Kind": 3,
                  "Straight": 4,
                  "Flush": 5,
                  "Full House": 6,
                  "Four of a Kind": 7,
                  "Straight Flush": 8,
                  "Royal Flush": 9}
    
    def __init__(self, cards):
        
        self._hand = sorted(list(map(lambda x: Card(x), cards)))
        self._handValue = self.checkHand(self._hand)
        
    def __eq__(self, other):
        if self._handValue != other._handValue:
            return self._handValue == other._handValue
        
        return (self._hand[4] == other._hand[4] and 
                self._hand[3] == other._hand[3] and 
                self._hand[2] == other._hand[2] and 
                self._hand[1] == other._hand[1] and 
                self._hand[0] == other._hand[0])
    
    def __lt__(self, other):
        if self._handValue != other._handValue:
            return self.handValues[self._handValue] < self.handValues[other._handValue]
        
        if self._handValue in ["Straight", "Straight Flush", "Royal Flush"]:
            return self._hand[0] < other._hand[0]
        
        if self._handValue in ["Three of a Kind", "Full House", "Four of a Kind"]:
            return self._hand[2] < other._hand[2]
        
        if self._handValue == "Two Pair":
            if self._hand[3] == other._hand[3]:
                if self._hand[1] == other._hand[1]:
                    return (self._hand[4] < other._hand[4] or 
                            self._hand[2] < other._hand[2] or 
                            self._hand[0] < other._hand[0])
                return self._hand[1] < other._hand[1]
            return self._hand[3] < other._hand[3]
        
        if self._handValue in ["High Card", "Flush"]:
            for n in range(5):
                if self._hand[4 - n] != other._hand[4 - n]:
                    return self._hand[4 - n] < other._hand[4 - n]
        
        if self._hand[3] == self._hand[4] or self._hand[3] == self._hand[2]:
            selfPair = self._hand[3]
        else:
            selfPair = self._hand[1]
        if other._hand[3] == other._hand[4] or other._hand[3] == other._hand[2]:
            otherPair = other._hand[3]
        else:
            otherPair = other._hand[1]
        if selfPair != otherPair:
            return selfPair < otherPair
        for n in range(5):
            if self._hand[4 - n] != other._hand[4 - n]:
                return self._hand[4 - n] < other._hand[4 - n]
        
    def __gt__(self, other):
        if self._handValue != other._handValue:
            return self.handValues[self._handValue] > self.handValues[other._handValue]
        
        if self._handValue in ["Straight", "Straight Flush", "Royal Flush"]:
            return self._hand[0] > other._hand[0]
        
        if self._handValue in ["Three of a Kind", "Full House", "Four of a Kind"]:
            return self._hand[2] > other._hand[2]
        
        if self._handValue == "Two Pair":
            if self._hand[3] == other._hand[3]:
                if self._hand[1] == other._hand[1]:
                    return (self._hand[4] > other._hand[4] or 
                            self._hand[2] > other._hand[2] or 
                            self._hand[0] > other._hand[0])
                return self._hand[1] > other._hand[1]
            return self._hand[3] > other._hand[3]
        
        if self._handValue in ["High Card", "Flush"]:
            for n in range(5):
                if self._hand[4 - n] != other._hand[4 - n]:
                    return self._hand[4 - n] > other._hand[4 - n]
                
        if self._hand[3] == self._hand[4] or self._hand[3] == self._hand[2]:
            selfPair = self._hand[3]
        else:
            selfPair = self._hand[1]
        if other._hand[3] == other._hand[4] or other._hand[3] == other._hand[2]:
            otherPair = other._hand[3]
        else:
            otherPair = other._hand[1]
        if selfPair != otherPair:
            return selfPair > otherPair
        for n in range(5):
            if self._hand[4 - n] != other._hand[4 - n]:
                return self._hand[4 - n] > other._hand[4 - n]
        
    def __repr__(self):
        return self._handValue
        
    def getHand(self):
        return self._hand
    
    def checkHand(self, hand):
        c1 = hand[0]
        straightCount = 1
        
        isFlush = True
        isStraight = True
        
        for card in hand[1:]:
            if card.getSuit() != c1.getSuit():
                isFlush = False
            if card.getNumericalValue() != c1.getNumericalValue() + straightCount:
                isStraight = False
                
            straightCount += 1
            
        if isFlush:
            if isStraight:
                if c1.getValue() == "T":
                    return "Royal Flush"
                return "Straight Flush"
            return "Flush"
        if isStraight:
            return "Straight"
        
        c2 = self._hand[1]
        c3 = self._hand[2]
        c4 = self._hand[3]
        c5 = self._hand[4]
        
        if c2 == c1:
            if c2 == c3:
                if c2 == c4:
                    return "Four of a Kind"
                if c4 == c5:
                    return "Full House"
                return "Three of a Kind"
            if c4 == c3:
                if c4 == c5:
                    return "Full House"
                return "Two Pairs"
            if c4 == c5:
                return "Two Pairs"
            return "One Pair"
        if c2 == c3:
            if c2 == c4:
                if c2 == c5:
                    return "Four of a Kind"
                return "Three of a Kind"
            if c4 == c5:
                return "Two Pairs"
            return "One Pair"
        if c4 == c3:
            if c4 == c5:
                return "Three of a Kind"
            return "One Pair"
        if c4 == c5:
            return "One Pair"
        
        return "High Card"

pokerFile = open("ez0054.txt", "r")
pokerHands = []

for row in pokerFile:
    row = row.strip()
    row = row.split(" ")
    pokerHands.append([row[0:5], row[5:]])
    
p1Score = 0
p2Score = 0
tie = 0

for hands in pokerHands:
    p1Hand = Hand(hands[0])
    p2Hand = Hand(hands[1])
    print(p1Hand.getHand(), "/", p2Hand.getHand())
    print(p1Hand, "/", p2Hand)
    
    
    if p1Hand > p2Hand:
        p1Score += 1
        print("Player 1 wins")
    elif p2Hand > p1Hand:
        p2Score += 1
        print("Player 2 wins")
    elif p1Hand == p2Hand:
        tie += 1
        
    print("\n")

print(p1Score, p2Score, tie)

File: javascript/completed/euler0054.js
import fs from 'fs';
import _ from 'lodash';

const SUITS = {
  C: 'Club',
  D: 'Diamond',
  H: 'Heart',
  S: 'Spade',
};

const VALUES = {
  2: { value: 'Two', rank: 2 },
  3: { value: 'Three', rank: 3 },
  4: { value: 'Four', rank: 4 },
  5: { value: 'Five', rank: 5 },
  6: { value: 'Six', rank: 6 },
  7: { value: 'Seven', rank: 7 },
  8: { value: 'Eight', rank: 8 },
  9: { value: 'Nine', rank: 9 },
  T: { value: 'Ten', rank: 10 },
  J: { value: 'Jack', rank: 11 },
  Q: { value: 'Queen', rank: 12 },
  K: { value: 'King', rank: 13 },
  A: { value: 'Ace', rank: 14 },
};

const HIGH_CARD = 'High Card';
const ONE_PAIR = 'One Pair';
const TWO_PAIRS = 'Two Pairs';
const THREE_OF_A_KIND = 'Three of a Kind';
const STRAIGHT = 'Straight';
const FLUSH = 'Flush';
const FULL_HOUSE = 'Full House';
const FOUR_OF_A_KIND = 'Four of a Kind';
const STRAIGHT_FLUSH = 'Straight Flush';
const ROYAL_FLUSH = 'Royal Flush';

const HAND_RANKS = {
  [HIGH_CARD]: 1,
  [ONE_PAIR]: 2,
  [TWO_PAIRS]: 3,
  [THREE_OF_A_KIND]: 4,
  [STRAIGHT]: 5,
  [FLUSH]: 6,
  [FULL_HOUSE]: 7,
  [FOUR_OF_A_KIND]: 8,
  [STRAIGHT_FLUSH]: 9,
  [ROYAL_FLUSH]: 10,
};

class Card {
  constructor(card) {
    this.suit = SUITS[card[1]];
    this.value = VALUES[card[0]].value;
    this.rank = VALUES[card[0]].rank;
  }

  getSuit() {
    return this.suit;
  }
  getRank() {
    return this.rank;
  }
  getValue() {
    return this.value;
  }
}

class Hand {
  constructor(cards) {
    this.cards = cards.map((c) => new Card(c)).sort((a, b) => a.getRank() - b.getRank());
    this.determineHandValue();
  }

  determineHandValue() {
    const { cards } = this;
    const cardValues = cards.map((c) => c.getRank());
    const cardSuits = cards.map((c) => c.getSuit());

    let result = {
      hand: undefined,
      prio1: undefined,
      prio2: undefined,
      prio3: undefined,
      prio4: undefined,
      prio5: undefined,
    };

    if (cardValues.length === _.uniq(cardValues).length) {
      result.prio1 = cards[4].getRank();
      result.prio2 = cards[3].getRank();
      result.prio3 = cards[2].getRank();
      result.prio4 = cards[1].getRank();
      result.prio5 = cards[0].getRank();

      if (
        cards[0].getRank() + 1 === cards[1].getRank() &&
        cards[1].getRank() + 1 === cards[2].getRank() &&
        cards[2].getRank() + 1 === cards[3].getRank() &&
        cards[3].getRank() + 1 === cards[4].getRank()
      ) {
        if (_.uniq(cardSuits).length === 1) {
          if (cards[0].getValue() === 'Ten') {
            result.hand = ROYAL_FLUSH;
          } else {
            result.hand = STRAIGHT_FLUSH;
          }
        } else {
          result.hand = STRAIGHT;
        }
      } else {
        if (_.uniq(cardSuits).length === 1) {
          // 424
          result.hand = FLUSH;
        } else {
          result.hand = HIGH_CARD;
        }
      }
    } else {
      const grouped = _.groupBy(cards, (c) => c.getRank());

      const maybeFour = Object.values(grouped).filter((v) => v.length === 4);
      const maybeThree = Object.values(grouped).filter((v) => v.length === 3);
      const maybeTwo = Object.values(grouped).filter((v) => v.length === 2);
      const maybeOne = Object.values(grouped).filter((v) => v.length === 1);

      if (maybeFour.length) {
        result.hand = FOUR_OF_A_KIND;
        result.prio1 = maybeFour[0][0].getRank();
        result.prio2 = maybeOne[0][0].getRank();
      } else if (maybeThree.length) {
        result.prio1 = maybeThree[0][0].getRank();

        if (maybeTwo.length) {
          result.hand = FULL_HOUSE;
          result.prio2 = maybeTwo[0][0].getRank();
        } else {
          result.hand = THREE_OF_A_KIND;
          result.prio2 = maybeOne[1][0].getRank();
          result.prio3 = maybeOne[0][0].getRank();
        }
      } else {
        if (maybeTwo.length === 2) {
          result.hand = TWO_PAIRS;
          result.prio1 = maybeTwo[1][0].getRank();
          result.prio2 = maybeTwo[0][0].getRank();
          result.prio3 = maybeOne[0][0].getRank();
        } else {
          result.hand = ONE_PAIR;
          result.prio1 = maybeTwo[0][0].getRank();
          result.prio2 = maybeOne[2][0].getRank();
          result.prio3 = maybeOne[1][0].getRank();
          result.prio4 = maybeOne[0][0].getRank();
        }
      }
    }

    return result;
  }

  isOwnBetter(hand) {
    const thisHand = this.determineHandValue();
    const thatHand = hand.determineHandValue();

    if (thisHand.hand !== thatHand.hand) {
      return HAND_RANKS[thisHand.hand] > HAND_RANKS[thatHand.hand];
    }
    if (thisHand.prio1 !== thatHand.prio1) {
      return thisHand.prio1 > thatHand.prio1;
    }
    if (thisHand.prio2 !== thatHand.prio2) {
      return thisHand.prio2 > thatHand.prio2;
    }
    if (thisHand.prio3 !== thatHand.prio3) {
      return thisHand.prio3 > thatHand.prio3;
    }
    if (thisHand.prio4 !== thatHand.prio4) {
      return thisHand.prio4 > thatHand.prio4;
    }
    return thisHand.prio5 > thatHand.prio5;
  }
}

export function solve54() {
  const file = fs.readFileSync('./data/x0054.txt', {
    encoding: 'utf-8',
  });
  const lines = file.split('\n');

  let player1Wins = [];

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const allCards = line.split(' ');
    const hand1 = new Hand(allCards.slice(0, 5));
    const hand2 = new Hand(allCards.slice(5));

    if (hand1.isOwnBetter(hand2)) {
      player1Wins.push(i + 1);
    }
  }

  return player1Wins.length;
}
